neuraxle/runners.py

```diff
@@ -0,0 +1,277 @@
+                data_inputs = step.transform(data_inputs)
+
+        return data_inputs
+
+    def transform_steps_left_to_do_recursive(self, step, data_inputs):
+        """
+        After loading the last checkpoint, recursively transform a step that contains steps
+        :param data_inputs: the data input to transform on
+        :param step: pipeline step
+        :return: the pipeline itself
+        """
+        checkpoint, index = self.find_starting_step_and_data_inputs(
+            steps_as_tuple=step.steps_as_tuple,
+            data_inputs=data_inputs
+        )
+
+        checkpoint_data_input, checkpoint_expected_outputs = checkpoint
+        if checkpoint_data_input is not None:
+            steps_left_to_do_recursive = step.steps_as_tuple[index:]
+            return self.transform_steps_left_to_do(checkpoint_data_input, steps_left_to_do_recursive)
+
+        return step.transform(data_inputs)
+
+    def find_starting_step_and_data_inputs(self, steps_as_tuple, data_inputs, expected_outputs=None):
+        """
+        Find the starting step index, and its corresponding data inputs using the checkpoint steps.
+        If the checkpoint is inside another step, the starting step will be step that contains the checkpoint
+        :param steps_as_tuple: steps
+        :param expected_outputs: expected outputs to fit on
+        :param data_inputs: the data input to fit on
+        :return: tuple(tuple(data_inputs, expected_outputs), step_index) tuple for the starting step data inputs-outputs,
+         and the starting step index
+        """
+        new_data_inputs = data_inputs
+        new_expected_outputs = expected_outputs
+        new_starting_step_index = 0
+        checkpoint_new_data_inputs = None
+        checkpoint_new_expected_outputs = None
+
+        for index, (step_name, step) in enumerate(reversed(steps_as_tuple)):
+            if isinstance(step, BaseCheckpointStep):
+                checkpoint_new_data_inputs, checkpoint_new_expected_outputs = self.load_checkpoint(step)
+
+            if isinstance(step, TruncableSteps) or isinstance(step, MetaStepsMixin):
+                checkpoint_new_data_inputs, checkpoint_new_expected_outputs = self.load_checkpoint_recursive(
+                    steps_as_tuple=step.steps_as_tuple,
+                    data_inputs=data_inputs,
+                    expected_outputs=expected_outputs
+                )
+
+            loaded_checkpoint_without_expected_outputs = expected_outputs is None and \
+                                                         checkpoint_new_data_inputs is not None
+            loaded_checkpoint_with_expected_outputs = expected_outputs is not None and \
+                                                      checkpoint_new_data_inputs is not None and \
+                                                      checkpoint_new_expected_outputs is not None
+
+            if loaded_checkpoint_without_expected_outputs or loaded_checkpoint_with_expected_outputs:
```

Comment:
if True? (reading variable names)

Reply:
it is possible to not have loaded any checkpoint at all with or without expected_outputs. I don't see what you mean here ? @guillaume-chevalier

Reply:
I just mean the variable names reads as if this if would take the value True. It's that the `if something or not_something` sounds like `if True` in every case. But I didn't try to understand this `if` yet, I'm just not sure of what it does. But this is a minor comment, the main point is to fix the TDA which is crucial.

Reply:
@alexbrillant The only thing I'm not sure about for the TDA it's how to do it during a `.fit`, for the `.transform` it should be easy to slice the TruncableSteps.

Reply:
@guillaume-chevalier yeah, it is not trivial. I am still trying to figure out how to do this cleanly

Reply:
See [this comment](https://github.com/Neuraxio/Neuraxle/pull/36#discussion_r315879939) for what I have in mind.